src/poc/miner_poc_mgr.erl

```diff
@@ -0,0 +1,1001 @@
+            %% we likely are the `heir', so we'll get it back if this process dies
+            %% TODO make table handling better, maybe a list of tables and iterate over
+            case maps:find(tab1, Args) of
+                error ->
+                    ok;
+                {ok, Tab1} ->
+                    true = ets:give_away(Tab1, Pid, undefined)
+            end,
+            case maps:find(tab2, Args) of
+                error ->
+                    ok;
+                {ok, Tab2} ->
+                    true = ets:give_away(Tab2, Pid, undefined)
+            end,
+            {ok, Pid};
+        Other ->
+            Other
+    end.
+
+-spec make_ets_table() -> [atom()].
+make_ets_table() ->
+    Tab1 = ets:new(
+        ?KEYS,
+        [
+            named_table,
+            public,
+            {heir, self(), undefined}
+        ]
+    ),
+    [Tab1].
+
+-spec save_local_poc_keys(CurHeight :: non_neg_integer(), [keys()]) -> ok.
+save_local_poc_keys(CurHeight, KeyList) ->
+    %% these are the keys ( public & private ) generated by this validator
+    %% as part of submitting a new heartbeat
+    %% push each key set to ets with a hash of the public key as key
+    %% each new block we will then check if any of our cached keys made it into the block
+    %% and if so retrieve the private key for each
+    [
+        begin
+            #{public := PubKey} = Keys,
+            OnionKeyHash = crypto:hash(sha256, libp2p_crypto:pubkey_to_bin(PubKey)),
+            POCKeyRec = #poc_local_key_data{receive_height = CurHeight, keys = Keys},
+            catch cache_local_poc_key(OnionKeyHash, POCKeyRec)
+        end
+        || Keys <- KeyList
+    ],
+    ok.
+
+-spec cached_local_poc_key(poc_key()) -> {ok, cached_local_poc_key_type()} | false.
+cached_local_poc_key(ID) ->
+    case ets:lookup(?KEYS, ID) of
+        [Res] -> {ok, Res};
+        _ -> false
+    end.
+
+-spec active_pocs()->[local_poc()].
+active_pocs() ->
+    gen_server:call(?MODULE, {active_pocs}).
+
+-spec check_target(
+    Challengee :: libp2p_crypto:pubkey_bin(),
+    BlockHash :: binary(),
+    OnionKeyHash :: binary()
+) -> false | {true, binary()} | {error, any()}.
+check_target(Challengee, BlockHash, OnionKeyHash) ->
+    LocalPOC = e2qc:cache(
+                local_pocs,
+                OnionKeyHash,
+                30,
+                fun() -> ?MODULE:local_poc(OnionKeyHash) end
+    ),
+    Res =
+        case LocalPOC of
+            {error, not_found} ->
+                %% if the cache returns not found it could be the poc has not yet been initialized
+                %% so check if we have a cached local POC key.
+                %% These are added when a val HB is submitted by the local node
+                %% if such a key exists its an indication the POC may not yet have been initialized
+                %% OR the e2qc cache was called before the POC was initialised and it
+                %% has cached the {error, not_found} term
```

Comment:
Is the cache really so important here or can we just query rocks directly?

Reply:
i think the cache is important here,  this is likely to be one of the hardest hit of the added APIs and a potential bottleneck.